### PR TITLE
[menu] move specific template logic which deal with menu-item's checked mod from menu to common menu_mode

### LINKED ISSUE #1513: [menu] base template for block contains logic only for one mod.
https://github.com/bem/bem-components/commit/da75b8177b9aa73ab80e068a7d55727c6daf6803 introduced logic which demanded only by `menu_mode_radio` template.
It iterates over content to set properties `_firstItem` and `_checkedItems`. Iteration function has assumption that `menu` should contain only `menu-item` or `menu__group`. It very inconvinient when we try to put inside block `menu` another elements or blocks which will be replaced with `menu-item` in later processing.

I suppose two enhancements:
- Move that specific code under `mod('mode', 'radio')` matcher.
- Make content iteration function indiferrent to block's content:
  - add `ctx.content = [].concat(ctx.content)` instead of `if(!this.isArray(ctx.content)) throw Error`
  - check on content property existance (`else if(ctx.content) { // menu__group`) before recursive call

Additionally we can support case of resolving blocks to `menu-item` which cannot be catched by iteration function and add js behavior in init section of `menu_mode_radio`, and document that in such cases live initialization should be disabled. But I think it can and should be done at project level.

/cc @hcodes @tadatuta

### PR DESCRIPTION
**Pull request by Guria**
iterate through content only within all menu_mode_*
do it in `content` mode instead of `def`
use applyNext instead of setting global ctx properties

resolves #1513  #1516

### CODE REVIEW

**Comment by Guria** [comment]
Why js tests could fail if all templates tests are ok? JS wasn't touched in commit.

**Comment by Guria** [comment]
It seems that spec tests doesn't include menu_mode.bemhtml into test bundle while tmpl-specs works as expected.

File: common.blocks/menu/_mode/menu_mode.bemhtml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+block('menu')
+    .match(function(){
+        return this.ctx.mods['mode'];
```
Comment (tadatuta) [inline common.blocks/menu/_mode/menu_mode.bemhtml]: I think it should be like this:

``` js
return typeof this.mods.mode !== 'undefined';
```

**Comment by Guria** [comment]
cc @tadatuta @blond please help
I have no enough experience to understand why menu_mode doesn't included in specs tests

**Comment by Guria** [comment]
@tadatuta I have resolved issue with specs tests.
Hope this pr will be merged

**Pull request closed without merging**